### PR TITLE
Add random sampling algorithm D

### PR DESCRIPTION
**Pull request by mrrobot-2000**
#552

### CODE REVIEW

File: src/sampling.jl [hunk 1]
```diff
@@ -284,11 +299,98 @@ function seqsample_c!(rng::AbstractRNG, a::AbstractArray, x::AbstractArray)
...
 end
-seqsample_c!(a::AbstractArray, x::AbstractArray) = seqsample_c!(Random.GLOBAL_RNG, a, x)
+seqsample_c!(a::AbstractArray, x::AbstractArray) =
+    seqsample_c!(Random.GLOBAL_RNG, a, x)
 
 ## TODO: implement Algorithm D (page 716 - 717)
+"""
+    seqsample_d!([rng], a::AbstractArray, x::AbstractArray)
+
+Random subsequence sampling using algorithm C described in the following paper (page 716-17):
```
Comment (mschauer) [inline src/sampling.jl]:
```suggestion
Random subsequence sampling using algorithm D described in the following paper (page 716-17):
```

File: src/sampling.jl [hunk 2]
```diff
@@ -720,18 +886,18 @@ function efraimidis_aexpj_wsample_norep!(rng::AbstractRNG, a::AbstractArray,
...
         w < 0 && error("Negative weight found in weight vector at index $i")
         w > 0 || continue
         X -= w
         X <= 0 || continue
 
         # update priority queue
-        t = exp(-w/threshold)
-        pq[1] = (-w/log(t+rand(rng)*(1-t)) => i)
+        t = exp(-w / threshold)
+        pq[1] = (-w / log(t + rand(rng) * (1 - t)) => i)
```
Comment (mschauer) [inline src/sampling.jl]:
```suggestion
        pq[1] = (-w/log(t + rand(rng)*(1 - t)) => i)
```
Use spacing to indicate order of operations, no need to put spaces around high precedence operators. Applies elsewhere.

File: src/sampling.jl [hunk 3]
```diff
@@ -593,27 +741,34 @@ Reference: Efraimidis, P. S., Spirakis, P. G. "Weighted random sampling with a r
...
+)
     n = length(a)
     length(wv) == n || throw(DimensionMismatch("a and wv must be of same length (got $n and $(length(wv)))."))
     k = length(x)
 
     # calculate keys for all items
     keys = randexp(rng, n)
-    for i in 1:n
-        @inbounds keys[i] = wv.values[i]/keys[i]
+    for i = 1:n
```
Comment (mschauer) [inline src/sampling.jl]: Why change one acceptable form into another? That only creates friction.

File: src/sampling.jl (hunk 2 above)
Comment (mrrobot-2000) [reply src/sampling.jl]: I had the Format Code option in Juno IDE. I think that caused these changes. I'll fix this everywhere.

File: src/sampling.jl (hunk 3 above)
Comment (mrrobot-2000) [reply src/sampling.jl]: Sorry. Again Format Code option.

**Comment by mrrobot-2000** [comment]
I do not have much experience with git. Therefore made some errors and had to these many commits to set everything straight. I eliminated all unnecessary changes. I think it is now ready for a fresh review.

File: src/sampling.jl [hunk 4]
```diff
@@ -286,7 +286,91 @@ function seqsample_c!(rng::AbstractRNG, a::AbstractArray, x::AbstractArray)
...
+
+    i = 0
+    j = 0
+
+    vprime = exp(log(rand(rng)) / n)
+    q1 = N - n + 1
+    q2 = q1 / N
+    alpha = 1 / 13 # choose alpha value
+    threshold = alpha * n
+    X = nothing
```
Comment (mschauer) [inline src/sampling.jl]:
```suggestion
```

File: src/sampling.jl [hunk 5]
```diff
@@ -286,7 +286,91 @@ function seqsample_c!(rng::AbstractRNG, a::AbstractArray, x::AbstractArray)
...
+
+    while n > 1 && threshold < N
+
+        while true
+
```
Comment (mschauer) [inline src/sampling.jl]:
```suggestion

            local X
```

File: src/sampling.jl [hunk 6]
```diff
@@ -286,7 +286,91 @@ function seqsample_c!(rng::AbstractRNG, a::AbstractArray, x::AbstractArray)
...
+            top = N - 1
+
+            while top >= limit
+                y = y * top / bottom
+                bottom -= 1
+                top -= 1
+            end
+
+            if exp(log(y) / (n - 1)) < N / (N - X)
```
Comment (mschauer) [inline src/sampling.jl]:
```suggestion
            if log(y) < (n - 1)*(log(N) - log(N - X))
```

File: src/sampling.jl [hunk 7]
```diff
@@ -286,7 +286,91 @@ function seqsample_c!(rng::AbstractRNG, a::AbstractArray, x::AbstractArray)
...
+                vprime = exp(log(rand(rng)) / (n - 1))
+                break
+            end
+            vprime = exp(log(rand(rng)) / n)
+        end
+
+        @inbounds x[j+=1] = a[i+=s+1]
```
Comment (mschauer) [inline src/sampling.jl]: I am not smart enough for increments in index expressions ;-)

File: src/sampling.jl [hunk 8]
```diff
@@ -286,7 +286,91 @@ function seqsample_c!(rng::AbstractRNG, a::AbstractArray, x::AbstractArray)
...
+"""
+function seqsample_d!(rng::AbstractRNG, a::AbstractArray, x::AbstractArray)
+    N = length(a)
+    n = length(x)
+    n <= N || error("length(x) should not exceed length(a)")
+
+    i = 0
+    j = 0
+
+    vprime = exp(log(rand(rng)) / n)
```
Comment (mschauer) [inline src/sampling.jl]:
```suggestion
    vprime = exp(-randexp(rng)/n)
```

File: src/sampling.jl (hunk 6 above)
Comment (mschauer) [reply src/sampling.jl]: Not sure, maybe it is important that this condition is computed in exactly the same way

File: src/sampling.jl (hunk 7 above)
Comment (mrrobot-2000) [reply src/sampling.jl]: Does this seem fine?
```suggestion
         j += 1
         i += s+1
        @inbounds x[j] = a[i]
```

File: src/sampling.jl (hunk 8 above)
Comment (mrrobot-2000) [reply src/sampling.jl]: Are you sure about this change ?

File: src/sampling.jl (hunk 6 above)
Comment (mrrobot-2000) [reply src/sampling.jl]: I think this should be fine.

File: src/sampling.jl (hunk 8 above)
Comment (mschauer) [reply src/sampling.jl]: The inverse of the cdf of an exponential random variable `randexp()` is `x = –log(1–u)`. Therefore `-log(1-rand())` and `-log(rand())` have the same distribution as `randexp`, and `randexp` has the better resolution. Not sure what matters here.

File: src/sampling.jl [hunk 9]
```diff
@@ -286,7 +286,91 @@ function seqsample_c!(rng::AbstractRNG, a::AbstractArray, x::AbstractArray)
...
+
+        while true
+
+            while true
+                X = N * (1 - vprime)
+                s = trunc(Int, X)
+                if s < q1
+                    break
+                end
+                vprime = exp(log(rand(rng)) / n)
```
Comment (mrrobot-2000) [inline src/sampling.jl]:
```suggestion
                vprime = exp(l-randexp(rng)/n)
```

File: src/sampling.jl [hunk 10]
```diff
@@ -286,7 +286,91 @@ function seqsample_c!(rng::AbstractRNG, a::AbstractArray, x::AbstractArray)
...
+            top = N - 1
+
+            while top >= limit
+                y = y * top / bottom
+                bottom -= 1
+                top -= 1
+            end
+
+            if exp(log(y) / (n - 1)) < N / (N - X)
+                vprime = exp(log(rand(rng)) / (n - 1))
```
Comment (mrrobot-2000) [inline src/sampling.jl]:
```suggestion
                vprime = exp(-randexp(rng) / (n-1))
```

File: src/sampling.jl [hunk 11]
```diff
@@ -286,7 +286,91 @@ function seqsample_c!(rng::AbstractRNG, a::AbstractArray, x::AbstractArray)
...
+                break
+            end
+            vprime = exp(log(rand(rng)) / n)
```
Comment (mrrobot-2000) [inline src/sampling.jl]:
```suggestion
            vprime = exp(-randexp(rng)/n)
```

**Comment by mrrobot-2000** [comment]
@mschauer I have made the changes suggested in the code review.

**Comment by mrrobot-2000** [comment]
Hi @mschauer, Can you check the latest commits and let me know why the Travis Build is failing only on Julia 1.4 for a completely unrelated test ?

File: test/sampling.jl [hunk 12]
```diff
@@ -122,8 +122,7 @@ function check_sample_norep(a::AbstractArray, vrgn, ptol::Real; ordered::Bool=fa
 end
 
 import StatsBase: knuths_sample!, fisher_yates_sample!, self_avoid_sample!
-import StatsBase: seqsample_a!, seqsample_c!
-
+import StatsBase: seqsample_a!, seqsample_c!, seqsample_d!
```
Comment (nalimilan) [inline test/sampling.jl]:
```suggestion
import StatsBase: seqsample_a!, seqsample_c!, seqsample_d!

```

File: src/sampling.jl [hunk 13]
```diff
@@ -286,7 +286,93 @@ function seqsample_c!(rng::AbstractRNG, a::AbstractArray, x::AbstractArray)
...
+
+            if log(y) < (n - 1)*(log(N) - log(N - X))
+                vprime = exp(-randexp(rng) / (n-1))
+                break
+            end
+            vprime = exp(-randexp(rng)/n)
+        end
+
+         j += 1
+         i += s+1
```
Comment (nalimilan) [inline src/sampling.jl]:
```suggestion
        j += 1
        i += s+1
```

File: src/sampling.jl [hunk 14]
```diff
@@ -286,7 +286,93 @@ function seqsample_c!(rng::AbstractRNG, a::AbstractArray, x::AbstractArray)
...
+    j = 0
+
+    vprime = exp(-randexp(rng)/n)
+    q1 = N - n + 1
+    q2 = q1 / N
+    alpha = 1 / 13 # choose alpha value
+    threshold = alpha * n
+
+    while n > 1 && threshold < N
+
```
Comment (nalimilan) [inline src/sampling.jl]:
```suggestion
```

File: src/sampling.jl [hunk 15]
```diff
@@ -286,7 +286,93 @@ function seqsample_c!(rng::AbstractRNG, a::AbstractArray, x::AbstractArray)
...
+        while true
+
```
Comment (nalimilan) [inline src/sampling.jl]:
```suggestion
```

File: src/sampling.jl [hunk 16]
```diff
@@ -286,7 +286,93 @@ function seqsample_c!(rng::AbstractRNG, a::AbstractArray, x::AbstractArray)
...
+            local X
```
Comment (nalimilan) [inline src/sampling.jl]:
```suggestion
           local X
```

**Comment by mschauer** [comment]
Nice, thank you.

**Pull request merged**